napari_btrack/track.py

```diff
@@ -31,23 +148,281 @@ def run_tracker(objects, config_file_path):
+                            value=sigma, name=f"{parameter}_sigma", annotation=float
+                        )
+                    ]
+                )
+            if (
+                parameter == "hypotheses"
+            ):  # this list should be represented as a series of checkboxes
+                for choice in default_value:
+                    widgets.extend(
+                        [create_widget(value=True, name=choice, annotation=bool)]
+                    )
+            else:  # use napari default
+                widgets.extend(
+                    [
+                        create_widget(
+                            value=default_value,
+                            name=parameter,
+                            annotation=type(default_value),
+                        )
+                    ]
+                )
+    return widgets
+
+
+def _create_napari_specific_widgets(widgets: List[Widget]) -> None:
+    """
+    Add the widgets which interact with napari itself
+    """
+    widgets.append(create_widget(**html_label_widget("Segmentation")))
+    labels = []
+    if napari.current_viewer():
+        labels = [
+            layer
+            for layer in napari.current_viewer().layers
+            if isinstance(layer, napari.layers.Labels)
+        ]
+    segmentation_widget = create_widget(
+        name="segmentation",
+        annotation=napari.layers.Labels,
+        options=dict(
+            tooltip="Should be a Labels layer. Convert an Image to Labels by right-clicking on it in the layers list, and clicking on 'Convert to Labels'",  # noqa: E501
+            choices=labels,
+        ),
+    )
+    widgets.append(segmentation_widget)
+
+
+def _create_pydantic_default_widgets(
+    widgets: List[Widget], config: TrackerConfig
+) -> None:
+    """
+    Create the widgets which have a tracker config equivalent.
+    """
+    widgets.extend(
+        [create_widget(name="max_search_radius", value=config.max_search_radius)]
     )
-    def widget(
-        viewer: napari.Viewer,
-        segmentation: napari.layers.Image,
-        config_file_path: Optional[Path],
-        reset_button,
+    model_configs = [config.motion_model, config.object_model, config.hypothesis_model]
+    model_widgets = [_create_per_model_widgets(model) for model in model_configs]
+    widgets.extend([item for sublist in model_widgets for item in sublist])
+
+
+def _create_cell_or_particle_widget(widgets: List[Widget]) -> None:
+    """Create a dropdown menu to choose between cell or particle mode."""
+    widgets.extend([create_widget(**html_label_widget("Mode"))])
+    widgets.extend(
+        [
+            create_widget(
+                name="mode", value="cell", options={"choices": ["cell", "particle"]}
+            )
+        ]
+    )
+
+
+def _widgets_to_tracker_config(container: Container) -> TrackerConfig:
+    """Helper function to convert from the widgets to a tracker configuration."""
+    motion_model_dict: Dict[str, Any] = {}
```

Comment:
Dict keys can also [be integers or other hashable things](https://wiki.python.org/moin/DictionaryKeys) - (I only consciously realised this because of this comment :flushed: - so thanks!)

Reply:
Ah good point!